src/ngraph/frontend/onnxifi/backend.cpp

```diff
@@ -0,0 +1,295 @@
+#include <cstring>
+
+#include <onnxifi.h>
+
+#include "backend.hpp"
+
+namespace ngraph
+{
+    namespace onnxifi
+    {
+        void Backend::get_capabilities(void* infoValue, std::size_t* infoValueSize) const
+        {
+            if (infoValueSize == nullptr)
+            {
+                throw std::invalid_argument{"null pointer"};
+            }
+            std::size_t requested{*infoValueSize};
+            *infoValueSize = sizeof(::onnxEnum);
+            if (requested < *infoValueSize)
+            {
+                throw std::length_error{"not enough space"};
+            }
+            if (infoValue != nullptr)
+            {
+                *reinterpret_cast<::onnxEnum*>(infoValue) = ONNXIFI_CAPABILITY_THREAD_SAFE;
+            }
+        }
+
+        void Backend::get_device(void* infoValue, std::size_t* infoValueSize) const
+        {
+            const constexpr char name[] = "nGraph ";
+            if (infoValueSize == nullptr)
+            {
+                throw std::invalid_argument{"null pointer"};
+            }
+            std::size_t requested{*infoValueSize};
+            *infoValueSize = sizeof name + m_type.size();
+            if (requested < *infoValueSize)
+            {
+                throw std::length_error{"not enough space"};
+            }
+            if (infoValue != nullptr)
+            {
+                std::memcpy(infoValue, name, sizeof name);
+                std::memcpy(
+                    reinterpret_cast<char*>(infoValue) + sizeof name, m_type.data(), m_type.size());
+            }
+        }
+
+        void Backend::get_device_type(void* infoValue, std::size_t* infoValueSize) const
+        {
+            if (infoValueSize == nullptr)
+            {
+                throw std::invalid_argument{"null pointer"};
+            }
+            std::size_t requested{*infoValueSize};
+            *infoValueSize = sizeof(::onnxEnum);
+            if (requested < *infoValueSize)
+            {
+                throw std::length_error{"not enough space"};
+            }
+            if (infoValue != nullptr)
+            {
+                if ((m_type == "CPU") || (m_type == "INTERPRETER"))
+                {
+                    *reinterpret_cast<::onnxEnum*>(infoValue) = ONNXIFI_DEVICE_TYPE_CPU;
+                }
+                else if ((m_type == "GPU") || (m_type == "INTELGPU"))
+                {
+                    *reinterpret_cast<::onnxEnum*>(infoValue) = ONNXIFI_DEVICE_TYPE_GPU;
+                }
+                else if (m_type == "FPGA")
+                {
+                    *reinterpret_cast<::onnxEnum*>(infoValue) = ONNXIFI_DEVICE_TYPE_FPGA;
+                }
+                else if (m_type == "NNP")
+                {
+                    *reinterpret_cast<::onnxEnum*>(infoValue) = ONNXIFI_DEVICE_TYPE_NPU;
+                }
+                else
+                {
+                    *reinterpret_cast<::onnxEnum*>(infoValue) = ONNXIFI_DEVICE_TYPE_HETEROGENEOUS;
```

Comment:
There is no 'unknown' or 'error' type for the device. Since "CPU" is the default backend for nGraph I may change this to default in case of ONNXIFI, too. Do we support heterogenous backends? What is the name for such a backend?

Reply:
We don't support heterogeneous as far as I know